commandapi-platforms/commandapi-paper/commandapi-paper-core/src/main/java/dev/jorel/commandapi/arguments/CustomArgument.java

```diff
@@ -0,0 +1,377 @@
+		}
+	}
+
+	/**
+	 * MessageBuilder is used to create error messages for invalid argument inputs
+	 */
+	public static class MessageBuilder implements Serializable {
+
+		private static final long serialVersionUID = 838497662821791798L;
+
+		StringBuilder builder;
+
+		/**
+		 * Create a blank message
+		 */
+		public MessageBuilder() {
+			builder = new StringBuilder();
+		}
+
+		/**
+		 * Create a message with an input string
+		 * 
+		 * @param str The string to start the message with
+		 */
+		public MessageBuilder(String str) {
+			builder = new StringBuilder(str);
+		}
+
+		/**
+		 * Appends the argument input that the CommandSender used in this command.<br>
+		 * For example, if <code>/foo bar</code> was executed and an error occurs with
+		 * the CustomArgument <code>bar</code>, then the arg input will append
+		 * <code>bar</code> to the end of the message.<br>
+		 * <br>
+		 * This input is determined at runtime, and is stored as <code>%input%</code>
+		 * until executed
+		 * 
+		 * @return A reference to this object
+		 */
+		public MessageBuilder appendArgInput() {
+			builder.append(INPUT);
+			return this;
+		}
+
+		/**
+		 * Appends the whole input that the CommandSender used in this command.<br>
+		 * For example, if <code>/foo bar</code> was executed, then <code>foo bar</code>
+		 * will be appended to the end of the message.<br>
+		 * <br>
+		 * This input is determined at runtime, and is stored as <code>%finput%</code>
+		 * until executed
+		 * 
+		 * @return A reference to this object
+		 */
+		public MessageBuilder appendFullInput() {
+			builder.append(FULL_INPUT);
+			return this;
+		}
+
+		/**
+		 * Appends <code>&lt;--[HERE]</code> to the end of the message
+		 * 
+		 * @return A reference to this object
+		 */
+		public MessageBuilder appendHere() {
+			builder.append("<--[HERE]");
+			return this;
+		}
+
+		/**
+		 * Appends a string to the end of this message
+		 * 
+		 * @param str The string to append to the end of this message
+		 * @return A reference to this object
+		 */
+		public MessageBuilder append(String str) {
+			builder.append(str);
+			return this;
+		}
+
+		/**
+		 * Appends an object to the end of this message
+		 * 
+		 * @param obj The object to append to the end of this message
+		 * @return A reference to this object
+		 */
+		public MessageBuilder append(Object obj) {
+			builder.append(obj);
+			return this;
+		}
+
+		/**
+		 * Returns the String content of this MessageBuilder
+		 * 
+		 * @return the String content of this MessageBuilder
+		 */
+		@Override
+		public String toString() {
+			return builder.toString();
+		}
+	}
+
+	/**
+	 * An exception used to create command-related errors for the CustomArgument
+	 */
+	@SuppressWarnings("serial")
+	public static class CustomArgumentException extends Exception {
+
+		private Component errorComponent = null;
+		private String errorMessage = null;
+		private MessageBuilder errorMessageBuilder = null;
```

Comment:
Perhaps instead of copying the whole `CustomArgument` class, which is otherwise identical Spigot/Paper, the inner `CustomArgumentException` class could be extracted and given the platform-specific component-type error messages.